File: 110-balanced-binary-tree/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
var dfs = function (node) {
    if (node === null) return [true, 0];
    let rightNode = dfs(node.right);
    let leftNode = dfs(node.left)
    return [rightNode[0] && leftNode[0] && Math.abs(leftNode[1] - rightNode[1]) <= 1, Math.max(rightNode[1], leftNode[1])+1];
}
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    return dfs(root)[0];
};

File: 110-balanced-binary-tree/balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    return true if root.nil?
    check_height(root) != -1
end


def check_height(node)
  return 0 if node.nil?

  left_height = check_height(node.left)
  return -1 if left_height == -1

  # Recursively check right subtree
  right_height = check_height(node.right)
  return -1 if right_height == -1

  # If current node is unbalanced, return -1
  if (left_height - right_height).abs > 1
    return -1
  end

  # Return actual height if balanced
  [left_height, right_height].max + 1
end